[ML] ECS Grok patterns in the _text_structure/find_structure endpoint

edsavage opened the pull request:
Add support for ECS compatible Grok patterns to the _text_structure/find_structure endpoint

Also add support for new CATALINA/TOMCAT timestamp formats used by ECS Grok patterns

Relates #77065

edsavage committed:
[ML] Add support for ECS compatible Grok patterns to the _text_structure/find_structure endpoint

Also add support for new CATALINA/TOMCAT timestamp formats used by ECS Grok patterns

Relates #77065

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

benwtrent commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/action/FindStructureAction.java:
```diff
@@ -141,6 +149,7 @@ public Request(StreamInput in) throws IOException {
             quote = in.readBoolean() ? (char) in.readVInt() : null;
             shouldTrimFields = in.readOptionalBoolean();
             grokPattern = in.readOptionalString();
+            ecsCompatibility = in.readOptionalString();
```
needs to be bwc safe for serialization between nodes.

benwtrent commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/action/FindStructureAction.java:
```diff
@@ -378,6 +418,7 @@ public void writeTo(StreamOutput out) throws IOException {
             }
             out.writeOptionalBoolean(shouldTrimFields);
             out.writeOptionalString(grokPattern);
+            out.writeOptionalString(ecsCompatibility);
```
needs to be bwc safe for serialization between nodes.

benwtrent commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java:
```diff
@@ -213,6 +221,7 @@ public TextStructure(StreamInput in) throws IOException {
         quote = in.readBoolean() ? (char) in.readVInt() : null;
         shouldTrimFields = in.readOptionalBoolean();
         grokPattern = in.readOptionalString();
+        ecsCompatibility = in.readOptionalString();
```
needs to be bwc safe for serialization between nodes.

benwtrent commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java:
```diff
@@ -254,6 +263,7 @@ public void writeTo(StreamOutput out) throws IOException {
         }
         out.writeOptionalBoolean(shouldTrimFields);
         out.writeOptionalString(grokPattern);
+        out.writeOptionalString(ecsCompatibility);
```
needs to be bwc safe for serialization between nodes.

benwtrent commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TextStructureUtils.java:
```diff
@@ -29,6 +29,9 @@
 public final class TextStructureUtils {
 
+    private static final String ECS_COMPATIBILITY_V1 = "v1";
+    private static final boolean ECS_COMPATIBILITY = false;
```
the name should indicate if its a default value or not.

benwtrent commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,182 @@ public final class TimestampFormatFinder {
         )
     );
 
+    /**
+     * The first match in this list will be chosen, so it needs to be ordered
+     * such that more generic patterns come after more specific patterns.
+     */
+    static final List<CandidateTimestampFormat> ORDERED_CANDIDATE_FORMATS_ECS_V1 = Arrays.asList(
```
From what I can tell, this list is EXACTLY the same as the non-ecs one except some new timestamps for catalina are prepended. 

Do we really need a whole new list here?

edsavage committed:
Attend to code review comments

edsavage committed:
Formatting

benwtrent reviewed (approved):
The changes as they are make sense to me. 

Would be good to get @droberts195 to take a look as well.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/action/FindStructureAction.java:
```diff
@@ -337,7 +373,16 @@ public ActionRequestValidationException validate() {
                         validationException
                     );
                 }
+                if (ECS_COMPATIBILITY_DISABLED.equalsIgnoreCase(ecsCompatibility) == false) {
+                    validationException = addIncompatibleArgErrorWithValue(
+                        ECS_COMPATIBILITY,
+                        TextStructure.Format.SEMI_STRUCTURED_TEXT,
+                        validationException,
+                        ecsCompatibility
+                    );
+                }
```
I think the validation on the request should be changed.

`ecs_compatibility` should be validated to be either not specified, `disabled`, or `v1` regardless of what the structure is.  At the moment we'll happily let through an invalid value if semi-structured text has been specified, which will then throw an exception later after other work has been done.  And it's quite possible we'll want to use Grok patterns as part of the analysis of more structured formats, so people should be able to tell us to use ECS patterns for those.

(The reason the overall `grok_pattern` is disallowed is that it applies to the entire message, not subsections of the analysis, so that really only makes sense for semi-structured logs.)

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into switch_to_ecs_grok_patterns

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+        // CATALINA7_DATESTAMP %{MONTH} %{MONTHDAY}, %{YEAR} %{HOUR}:%{MINUTE}:%{SECOND} (?:AM|PM)
+        items.add(
+            new CandidateTimestampFormat(
+                example -> Collections.singletonList("MMM dd, yyyy hh:mm:ss a"),
```
```suggestion
                example -> Collections.singletonList("MMM dd, yyyy h:mm:ss a"),
```

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+        items.add(
+            new CandidateTimestampFormat(
+                example -> Collections.singletonList("MMM dd, yyyy hh:mm:ss a"),
+                "\\b[A-Z]\\S{2} \\d{2}, \\d{2} \\d{4}:\\d{2}:\\d{2}[:.,]\\d{3}",
```
```suggestion
                "\\b[A-Z]\\S{2} \\d{2}, \\d{4} \\d{1,2}:\\d{2}:\\d{2} [AP]M",
```

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+    static {
+        List<CandidateTimestampFormat> items = new ArrayList<>();
+        // CATALINA8_DATESTAMP %{MONTHDAY}-%{MONTH}-%{YEAR} %{HOUR}:%{MINUTE}:%{SECOND}
+        // 29-Aug-2009 12:03:33
```
```suggestion
        // 29-Aug-2021 12:03:33.578
```

According to the example in the pattern file this always has millisecond accuracy.

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+                "\\b\\d{2}-[A-Z]\\S{2}-\\d{4} \\d{2}:\\d{2}:\\d{2}[:.,]\\d{3}",
+                "\\b%{MONTHDAY}-%{MONTH}-%{YEAR} %{HOUR}:%{MINUTE}:%{SECOND}\\b",
+                "CATALINA8_DATESTAMP",
+                "11     1111 11 11 11",
```
```suggestion
                "11     1111 11 11 11 111",
```

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+                "\\b[A-Z]\\S{2} \\d{2}, \\d{2} \\d{4}:\\d{2}:\\d{2}[:.,]\\d{3}",
+                "\\b%{MONTH} %{MONTHDAY}, %{YEAR} %{HOUR}:%{MINUTE}:%{SECOND} (?:AM|PM)\\b",
+                "CATALINA7_DATESTAMP",
+                "    11  1111 11 11 11 111",
```
```suggestion
                Arrays.asList("    11  1111 1 11 11", "    11  1111 11 11 11"),,
```

This one is basically the old `CATALINA_DATESTAMP`, but with the hardcoded `20` in the year replaced.

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+                "CATALINA7_DATESTAMP",
+                "    11  1111 11 11 11 111",
+                0,
+                13
```
```suggestion
                3
```

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+                "CATALINA8_DATESTAMP",
+                "11     1111 11 11 11",
+                0,
+                13
```
```suggestion
                0
```

edsavage committed:
Attend to review comments

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into switch_to_ecs_grok_patterns

edsavage committed:
Attend to further review comments

edsavage committed:
Typo.. sigh

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+        // 29-Aug-2021 12:03:33.578
+        items.add(
+            new CandidateTimestampFormat(
+                example -> Collections.singletonList("dd-MMM-yyyy hh:mm:ss"),
+                "\\b\\d{2}-[A-Z]\\S{2}-\\d{4} \\d{2}:\\d{2}:\\d{2}[:.,]\\d{3}",
```
If the comment is accurate and the separator is always `.` then this should be:

```suggestion
                example -> Collections.singletonList("dd-MMM-yyyy hh:mm:ss.SSS"),
                "\\b\\d{2}-[A-Z]\\S{2}-\\d{4} \\d{2}:\\d{2}:\\d{2}\\.\\d{3}",
```

Otherwise, if the fractional seconds separator really can vary between `.`, `,` and `:` even for this single type of logs then line 303 would need to be `example -> CandidateTimestampFormat.adjustFractionalSecondsFromEndOfExample("dd-MMM-yyyy hh:mm:ss"),` and line 304 can stay as-is.

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TimestampFormatFinder.java:
```diff
@@ -288,6 +289,43 @@ public final class TimestampFormatFinder {
+                3
+            )
+        );
+        items.addAll(ORDERED_CANDIDATE_FORMATS);
```
Maybe this should be "add all except `CATALINA_DATESTAMP`" (because that's been replaced by the two version-specific formats).

edsavage committed:
Adjustments for CATALINA datestamps

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/action/FindStructureAction.java:
```diff
@@ -262,6 +274,16 @@ public void setGrokPattern(String grokPattern) {
+            return ecsCompatibility;
+        }
+
+        public void setEcsCompatibility(String ecsCompatibility) {
+            this.ecsCompatibility = (ecsCompatibility == null || ecsCompatibility.isEmpty())
+                ? ECS_COMPATIBILITY_DISABLED
+                : ecsCompatibility;
+        }
```
This is inconsistent with other code that sets `ecsCompatibility`.  The other code, for example the parser and both constructors, allow it to be `null`.  This method converts `null` to `"disabled"`.  It should be handled consistently so that either everywhere where `ecsCompatibility` is set uses `disabled` instead of `null` or else this setting mechanism considers `null` a separate value.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/action/FindStructureAction.java:
```diff
@@ -338,6 +360,22 @@ public ActionRequestValidationException validate() {
+                validationException = addValidationError(
+                    "["
+                        + ECS_COMPATIBILITY.getPreferredName()
+                        + "] must either ["
+                        + ECS_COMPATIBILITY_V1
+                        + "] or ["
+                        + ECS_COMPATIBILITY_DISABLED
+                        + "] if specified",
```
```suggestion
                        + "] must be one of ["
                        + String.join(", ", Grok.ECS_COMPATIBILITY_MODES)
                        + "] if specified",
```

droberts195 commented on x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/structurefinder/TextStructureUtils.java:
```diff
@@ -494,6 +529,7 @@ public static Map<String, Object> makeIngestPipelineDefinition(
             if (customGrokPatternDefinitions.isEmpty() == false) {
                 grokProcessorSettings.put("pattern_definitions", customGrokPatternDefinitions);
             }
+            grokProcessorSettings.put("ecs_compatibility", ecsCompatibility);
```
It is possible for `ecsCompatibility` to be `null` here, because `TextStructureOverrides` doesn't enforce that `ecsCompatibility` is set to something.  I think it is actually good if we explicitly set it here even if the user-supplied overrides didn't set it, because then the ingest pipeline we create will not change meaning if the default value of `ecs_compatibility` is changed in a future major release.

It would be good to think through how consistency in the values of `ecsCompatibility` can be enforced.  Maybe every class that stores it should default it to `disabled` and then enforce that it is never set to `null` after that.  Then the `writeOptionalString` and `readOptionalString` methods used with it become `writeString` and `readString`.

And this line can be

```suggestion
            grokProcessorSettings.put("ecs_compatibility", Objects.requireNonNull(ecsCompatibility));
```

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/action/FindStructureAction.java:
```diff
@@ -30,6 +32,8 @@
 import static org.elasticsearch.action.ValidateActions.addValidationError;
 
 public class FindStructureAction extends ActionType<FindStructureAction.Response> {
+    public static final String ECS_COMPATIBILITY_DISABLED = "disabled";
+    public static final String ECS_COMPATIBILITY_V1 = "v1";
```
```suggestion
    public static final String ECS_COMPATIBILITY_DISABLED = Grok.ECS_COMPATIBILITY_MODES[0];
    public static final String ECS_COMPATIBILITY_V1 = Grok.ECS_COMPATIBILITY_MODES[1];
```

droberts195 reviewed (commented):
The YAML tests for this endpoint don't handle unstructured text well, so there's no coverage for it.

Please can you run some manual tests using `./gradlew run` and `curl`.

For example:

```
curl -u elastic:password -s -H "Content-Type: application/json" -XPOST "localhost:9200/_text_structure/find_structure?pretty&explain" -T my_test_file
curl -u elastic:password -s -H "Content-Type: application/json" -XPOST "localhost:9200/_text_structure/find_structure?pretty&explain&ecs_compatibility=disabled" -T my_test_file
curl -u elastic:password -s -H "Content-Type: application/json" -XPOST "localhost:9200/_text_structure/find_structure?pretty&explain&ecs_compatibility=v1" -T my_test_file
```

And confirm you get sensible output in all 3 cases for a few different log files.

edsavage committed:
Deal consistently with ecsCompatibility field being null

edsavage commented:
The outcomes of manual testing using `curl` do return sensible results, e.g, with an input file `raw_messages.log` with contents:

```
head raw_messages.log 
223.87.60.27 - - [2018-07-22T00:39:02.912Z] "GET /elasticsearch/elasticsearch-6.3.2.deb_1 HTTP/1.1" 200 6219 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1"
130.246.123.197 - - [2018-07-22T03:26:21.326Z] "GET /beats/metricbeat_1 HTTP/1.1" 200 6850 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1"
120.49.143.213 - - [2018-07-22T03:30:25.131Z] "GET /styles/main.css_1 HTTP/1.1" 503 0 "-" "Mozilla/5.0 (X11; Linux i686) AppleWebKit/534.24 (KHTML, like Gecko) Chrome/11.0.696.50 Safari/534.24"
99.74.118.237 - - [2018-07-22T03:34:43.399Z] "GET /beats/metricbeat/metricbeat-6.3.2-amd64.deb_1 HTTP/1.1" 200 14113 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; .NET CLR 1.1.4322)"
177.111.217.54 - - [2018-07-22T03:37:04.863Z] "GET /enterprise_1 HTTP/1.1" 200 2492 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; .NET CLR 1.1.4322)"
106.225.58.146 - - [2018-07-22T03:49:40.669Z] "GET /apm_1 HTTP/1.1" 503 0 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1"
6.138.148.165 - - [2018-07-22T03:57:39.612Z] "GET /beats/metricbeat/metricbeat-6.3.2-amd64.deb_1 HTTP/1.1" 200 1872 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1"
218.148.135.12 - - [2018-07-22T04:18:12.345Z] "GET /beats/filebeat/filebeat-6.3.2-linux-x86_64.tar.gz_1 HTTP/1.1" 200 4531 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1"
216.9.22.134 - - [2018-07-22T05:27:11.939Z] "GET /beats/metricbeat_1 HTTP/1.1" 200 3629 "-" "Mozilla/5.0 (X11; Linux i686) AppleWebKit/534.24 (KHTML, like Gecko) Chrome/11.0.696.50 Safari/534.24"
12.132.31.17 - - [2018-07-22T05:36:25.812Z] "GET /elasticsearch/elasticsearch-6.3.2.tar.gz_1 HTTP/1.1" 200 9797 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1"
```
Gives the following results for various values of `ecs_compatibility`

1. ecs_compatibility=v1

```
curl --insecure -u elastic:password -s -H "Content-Type: application/json" -XPOST "https://localhost:9200/_text_structure/find_structure?pretty&explain&ecs_compatibility=v1" -T raw_messages.log
{
  "num_lines_analyzed" : 999,
  "num_messages_analyzed" : 999,
  "sample_start" : "223.87.60.27 - - [2018-07-22T00:39:02.912Z] \"GET /elasticsearch/elasticsearch-6.3.2.deb_1 HTTP/1.1\" 200 6219 \"-\" \"Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1\"\n130.246.123.197 - - [2018-07-22T03:26:21.326Z] \"GET /beats/metricbeat_1 HTTP/1.1\" 200 6850 \"-\" \"Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1\"\n",
  "charset" : "UTF-8",
  "has_byte_order_marker" : false,
  "format" : "semi_structured_text",
  "multiline_start_pattern" : "^.*? .*? .*? \\[\\b\\d{4}-\\d{2}-\\d{2}[T ]\\d{2}:\\d{2}",
  "grok_pattern" : "%{IP:ipaddress} .*? .*? \\[%{TIMESTAMP_ISO8601:timestamp}\\] \".*? %{PATH:path} .*?/%{NUMBER:field2}%{QUOTEDSTRING:field}.*?%{QUOTEDSTRING:field3}.*?/%{NUMBER:field4} \\(.*?; .*? .*?%{NUMBER:field5}.*",
  "ecs_compatibility" : "v1",
  "timestamp_field" : "timestamp",
  "joda_timestamp_formats" : [
    "ISO8601"
  ],
  "java_timestamp_formats" : [
    "ISO8601"
  ],
  "need_client_timezone" : false,
  "mappings" : {
    "properties" : {
      "@timestamp" : {
        "type" : "date"
      },
      "field" : {
        "type" : "keyword"
      },
      "field2" : {
        "type" : "double"
      },
      "field3" : {
        "type" : "keyword"
      },
      "field4" : {
        "type" : "double"
      },
      "field5" : {
        "type" : "double"
      },
      "ipaddress" : {
        "type" : "ip"
      },
      "message" : {
        "type" : "text"
      },
      "path" : {
        "type" : "keyword"
      }
    }
  },
  "ingest_pipeline" : {
    "description" : "Ingest pipeline created by text structure finder",
    "processors" : [
      {
        "grok" : {
          "field" : "message",
          "patterns" : [
            "%{IP:ipaddress} .*? .*? \\[%{TIMESTAMP_ISO8601:timestamp}\\] \".*? %{PATH:path} .*?/%{NUMBER:field2}%{QUOTEDSTRING:field}.*?%{QUOTEDSTRING:field3}.*?/%{NUMBER:field4} \\(.*?; .*? .*?%{NUMBER:field5}.*"
          ],
          "ecs_compatibility" : "v1"
        }
      },
```

2. ecs_compatibility=disabled

```
curl --insecure -u elastic:password -s -H "Content-Type: application/json" -XPOST "https://localhost:9200/_text_structure/find_structure?pretty&explain&ecs_compatibility=disabled" -T raw_messages.log
{
  "num_lines_analyzed" : 999,
  "num_messages_analyzed" : 999,
  "sample_start" : "223.87.60.27 - - [2018-07-22T00:39:02.912Z] \"GET /elasticsearch/elasticsearch-6.3.2.deb_1 HTTP/1.1\" 200 6219 \"-\" \"Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1\"\n130.246.123.197 - - [2018-07-22T03:26:21.326Z] \"GET /beats/metricbeat_1 HTTP/1.1\" 200 6850 \"-\" \"Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1\"\n",
  "charset" : "UTF-8",
  "has_byte_order_marker" : false,
  "format" : "semi_structured_text",
  "multiline_start_pattern" : "^.*? .*? .*? \\[\\b\\d{4}-\\d{2}-\\d{2}[T ]\\d{2}:\\d{2}",
  "grok_pattern" : "%{IP:ipaddress} .*? .*? \\[%{TIMESTAMP_ISO8601:timestamp}\\] \".*? %{PATH:path} .*?/%{NUMBER:field2}%{QUOTEDSTRING:field}.*?%{QUOTEDSTRING:field3}.*?/%{NUMBER:field4} \\(.*?; .*? .*?%{NUMBER:field5}.*",
  "ecs_compatibility" : "disabled",
  "timestamp_field" : "timestamp",
  "joda_timestamp_formats" : [
    "ISO8601"
  ],
  "java_timestamp_formats" : [
    "ISO8601"
  ],
  "need_client_timezone" : false,
  "mappings" : {
    "properties" : {
      "@timestamp" : {
        "type" : "date"
      },
      "field" : {
        "type" : "keyword"
      },
      "field2" : {
        "type" : "double"
      },
      "field3" : {
        "type" : "keyword"
      },
      "field4" : {
        "type" : "double"
      },
      "field5" : {
        "type" : "double"
      },
      "ipaddress" : {
        "type" : "ip"
      },
      "message" : {
        "type" : "text"
      },
      "path" : {
        "type" : "keyword"
      }
    }
  },
  "ingest_pipeline" : {
    "description" : "Ingest pipeline created by text structure finder",
    "processors" : [
      {
        "grok" : {
          "field" : "message",
          "patterns" : [
            "%{IP:ipaddress} .*? .*? \\[%{TIMESTAMP_ISO8601:timestamp}\\] \".*? %{PATH:path} .*?/%{NUMBER:field2}%{QUOTEDSTRING:field}.*?%{QUOTEDSTRING:field3}.*?/%{NUMBER:field4} \\(.*?; .*? .*?%{NUMBER:field5}.*"
          ],
          "ecs_compatibility" : "disabled"
        }
      },
```

3. ecs_compatibility not specified

```
curl --insecure -u elastic:password -s -H "Content-Type: application/json" -XPOST "https://localhost:9200/_text_structure/find_structure?pretty&explain" -T raw_messages.log 
{
  "num_lines_analyzed" : 999,
  "num_messages_analyzed" : 999,
  "sample_start" : "223.87.60.27 - - [2018-07-22T00:39:02.912Z] \"GET /elasticsearch/elasticsearch-6.3.2.deb_1 HTTP/1.1\" 200 6219 \"-\" \"Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1\"\n130.246.123.197 - - [2018-07-22T03:26:21.326Z] \"GET /beats/metricbeat_1 HTTP/1.1\" 200 6850 \"-\" \"Mozilla/5.0 (X11; Linux x86_64; rv:6.0a1) Gecko/20110421 Firefox/6.0a1\"\n",
  "charset" : "UTF-8",
  "has_byte_order_marker" : false,
  "format" : "semi_structured_text",
  "multiline_start_pattern" : "^.*? .*? .*? \\[\\b\\d{4}-\\d{2}-\\d{2}[T ]\\d{2}:\\d{2}",
  "grok_pattern" : "%{IP:ipaddress} .*? .*? \\[%{TIMESTAMP_ISO8601:timestamp}\\] \".*? %{PATH:path} .*?/%{NUMBER:field2}%{QUOTEDSTRING:field}.*?%{QUOTEDSTRING:field3}.*?/%{NUMBER:field4} \\(.*?; .*? .*?%{NUMBER:field5}.*",
  "ecs_compatibility" : "disabled",
  "timestamp_field" : "timestamp",
  "joda_timestamp_formats" : [
    "ISO8601"
  ],
  "java_timestamp_formats" : [
    "ISO8601"
  ],
  "need_client_timezone" : false,
  "mappings" : {
    "properties" : {
      "@timestamp" : {
        "type" : "date"
      },
      "field" : {
        "type" : "keyword"
      },
      "field2" : {
        "type" : "double"
      },
      "field3" : {
        "type" : "keyword"
      },
      "field4" : {
        "type" : "double"
      },
      "field5" : {
        "type" : "double"
      },
      "ipaddress" : {
        "type" : "ip"
      },
      "message" : {
        "type" : "text"
      },
      "path" : {
        "type" : "keyword"
      }
    }
  },
  "ingest_pipeline" : {
    "description" : "Ingest pipeline created by text structure finder",
    "processors" : [
      {
        "grok" : {
          "field" : "message",
          "patterns" : [
            "%{IP:ipaddress} .*? .*? \\[%{TIMESTAMP_ISO8601:timestamp}\\] \".*? %{PATH:path} .*?/%{NUMBER:field2}%{QUOTEDSTRING:field}.*?%{QUOTEDSTRING:field3}.*?/%{NUMBER:field4} \\(.*?; .*? .*?%{NUMBER:field5}.*"
          ],
          "ecs_compatibility" : "disabled"
        }
      },
```

droberts195 commented:
In the example you tested the ECS and non-ECS Grok patterns were the same.  It would be interesting to test one where they'll be different, like a Catalina 7 log file.  Try this:

```
Oct 09, 2013 12:21:37 PM org.apache.catalina.core.AprLifecycleListener init
INFO: The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: /usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib
Oct 09, 2013 12:21:37 PM org.apache.coyote.http11.Http11Protocol init
INFO: Initializing Coyote HTTP/1.1 on http-8080
Oct 09, 2013 12:21:37 PM org.apache.catalina.startup.Catalina load
INFO: Initialization processed in 461 ms
Oct 09, 2013 12:21:37 PM org.apache.catalina.core.StandardService start
INFO: Starting service Catalina
Oct 09, 2013 12:21:37 PM org.apache.catalina.core.StandardEngine start
INFO: Starting Servlet Engine: Apache Tomcat/6.0.29
Oct 09, 2013 12:21:37 PM org.apache.catalina.startup.HostConfig deployDescriptor
INFO: Deploying configuration descriptor manager.xml
Oct 09, 2013 12:21:37 PM org.apache.catalina.startup.HostConfig deployDescriptor
INFO: Deploying configuration descriptor host-manager.xml
Oct 09, 2013 12:21:39 PM org.apache.catalina.startup.HostConfig deployWAR
INFO: Deploying web application archive prelert.war
```

edsavage commented:
Testing with CATALINA7 format log messages:

1. With ecs_compatibility disabled
```
curl --insecure -u elastic:OT_FIBhd-uR2=tX4Wyq8 -s -H "Content-Type: application/json" -XPOST "https://localhost:9200/_text_structure/find_structure?pretty&explain&ecs_compatibility=disabled" -T catalina7.log
{
  "num_lines_analyzed" : 14,
  "num_messages_analyzed" : 7,
  "sample_start" : "Oct 09, 2013 12:21:37 PM org.apache.catalina.core.AprLifecycleListener init\nINFO: The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: /usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib\nOct 09, 2013 12:21:37 PM org.apache.coyote.http11.Http11Protocol init\nINFO: Initializing Coyote HTTP/1.1 on http-8080\n",
  "charset" : "UTF-8",
  "has_byte_order_marker" : false,
  "format" : "semi_structured_text",
  "multiline_start_pattern" : "^[A-Z]\\S{2} \\d{2}, \\d{4} \\d{1,2}:\\d{2}:\\d{2} [AP]M\\b",
  "grok_pattern" : "%{CATALINALOG}",
  "ecs_compatibility" : "disabled",
  "timestamp_field" : "timestamp",
  "joda_timestamp_formats" : [
    "MMM dd, YYYY h:mm:ss a"
  ],
  "java_timestamp_formats" : [
    "MMM dd, yyyy h:mm:ss a"
  ],
  "need_client_timezone" : true,
  "mappings" : {
    "properties" : {
      "@timestamp" : {
        "type" : "date"
      },
      "class" : {
        "type" : "keyword"
      },
      "logmessage" : {
        "type" : "keyword"
      },
      "message" : {
        "type" : "text"
      }
    }
  },
  "ingest_pipeline" : {
    "description" : "Ingest pipeline created by text structure finder",
    "processors" : [
      {
        "grok" : {
          "field" : "message",
          "patterns" : [
            "%{CATALINALOG}"
          ],
          "ecs_compatibility" : "disabled"
        }
      },
      {
        "date" : {
          "field" : "timestamp",
          "timezone" : "{{ event.timezone }}",
          "formats" : [
            "MMM dd, yyyy h:mm:ss a"
          ]
        }
      },
      {
        "remove" : {
          "field" : "timestamp"
        }
      }
    ]
  },
  "field_stats" : {
    "class" : {
      "count" : 7,
      "cardinality" : 6,
      "top_hits" : [
        {
          "value" : "org.apache.catalina.startup.HostConfig",
          "count" : 2
        },
        {
          "value" : "org.apache.catalina.core.AprLifecycleListener",
          "count" : 1
        },
        {
          "value" : "org.apache.catalina.core.StandardEngine",
          "count" : 1
        },
        {
          "value" : "org.apache.catalina.core.StandardService",
          "count" : 1
        },
        {
          "value" : "org.apache.catalina.startup.Catalina",
          "count" : 1
        },
        {
          "value" : "org.apache.coyote.http11.Http11Protocol",
          "count" : 1
        }
      ]
    },
    "logmessage" : {
      "count" : 7,
      "cardinality" : 4,
      "top_hits" : [
        {
          "value" : "deployDescriptor",
          "count" : 2
        },
        {
          "value" : "init",
          "count" : 2
        },
        {
          "value" : "start",
          "count" : 2
        },
        {
          "value" : "load",
          "count" : 1
        }
      ]
    },
    "message" : {
      "count" : 7,
      "cardinality" : 7,
      "top_hits" : [
        {
          "value" : "Oct 09, 2013 12:21:37 PM org.apache.catalina.core.AprLifecycleListener init\nINFO: The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: /usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib",
          "count" : 1
        },
        {
          "value" : "Oct 09, 2013 12:21:37 PM org.apache.catalina.core.StandardEngine start\nINFO: Starting Servlet Engine: Apache Tomcat/6.0.29",
          "count" : 1
        },
        {
          "value" : "Oct 09, 2013 12:21:37 PM org.apache.catalina.core.StandardService start\nINFO: Starting service Catalina",
          "count" : 1
        },
        {
          "value" : "Oct 09, 2013 12:21:37 PM org.apache.catalina.startup.Catalina load\nINFO: Initialization processed in 461 ms",
          "count" : 1
        },
        {
          "value" : "Oct 09, 2013 12:21:37 PM org.apache.catalina.startup.HostConfig deployDescriptor\nINFO: Deploying configuration descriptor host-manager.xml",
          "count" : 1
        },
        {
          "value" : "Oct 09, 2013 12:21:37 PM org.apache.catalina.startup.HostConfig deployDescriptor\nINFO: Deploying configuration descriptor manager.xml",
          "count" : 1
        },
        {
          "value" : "Oct 09, 2013 12:21:37 PM org.apache.coyote.http11.Http11Protocol init\nINFO: Initializing Coyote HTTP/1.1 on http-8080",
          "count" : 1
        }
      ]
    },
    "timestamp" : {
      "count" : 7,
      "cardinality" : 1,
      "earliest" : "Oct 09, 2013 12:21:37 PM",
      "latest" : "Oct 09, 2013 12:21:37 PM",
      "top_hits" : [
        {
          "value" : "Oct 09, 2013 12:21:37 PM",
          "count" : 7
        }
      ]
    }
  },
  "explanation" : [
    "Using character encoding [UTF-8], which matched the input with [15%] confidence - first [8kB] of input was pure ASCII",
    "Not NDJSON because there was a parsing exception: [[1:4] Unrecognized token 'Oct': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false') at [Source: (org.elasticsearch.xpack.textstructure.structurefinder.NdJsonTextStructureFinderFactory$ContextPrintingStringReader); line: 1, column: 4]]",
    "Not XML because there was a parsing exception: [ParseError at [row,col]:[1,1] Message: Content is not allowed in prolog.]",
    "Not CSV because 1 or more rows did not have the same number of fields as the first row (2). Bad rows [1]",
    "Not TSV because the first row has fewer than [2] fields: [1]",
    "Not semicolon delimited values because the first row has fewer than [4] fields: [1]",
    "Not vertical line delimited values because the first row has fewer than [5] fields: [1]",
    "Deciding sample is text",
    "Most likely timestamp format is [MMM dd, yyyy h:mm:ss a]",
    "A full message Grok pattern [CATALINALOG] looks appropriate"
  ]
}
```

2. With ecs_compatibility v1
```
curl --insecure -u elastic:OT_FIBhd-uR2=tX4Wyq8 -s -H "Content-Type: application/json" -XPOST "https://localhost:9200/_text_structure/find_structure?pretty&explain&ecs_compatibility=v1" -T catalina7.log
{
  "num_lines_analyzed" : 14,
  "num_messages_analyzed" : 7,
  "sample_start" : "Oct 09, 2013 12:21:37 PM org.apache.catalina.core.AprLifecycleListener init\nINFO: The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: /usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib\nOct 09, 2013 12:21:37 PM org.apache.coyote.http11.Http11Protocol init\nINFO: Initializing Coyote HTTP/1.1 on http-8080\n",
  "charset" : "UTF-8",
  "has_byte_order_marker" : false,
  "format" : "semi_structured_text",
  "multiline_start_pattern" : "^[A-Z]\\S{2} \\d{2}, \\d{4} \\d{1,2}:\\d{2}:\\d{2} [AP]M\\b",
  "grok_pattern" : "%{CATALINALOG}",
  "ecs_compatibility" : "v1",
  "timestamp_field" : "timestamp",
  "joda_timestamp_formats" : [
    "MMM dd, YYYY h:mm:ss a"
  ],
  "java_timestamp_formats" : [
    "MMM dd, yyyy h:mm:ss a"
  ],
  "need_client_timezone" : true,
  "mappings" : {
    "properties" : {
      "@timestamp" : {
        "type" : "date"
      },
      "java.log.origin.class.name" : {
        "type" : "keyword"
      },
      "log.level" : {
        "type" : "keyword"
      },
      "log.origin.function" : {
        "type" : "keyword"
      },
      "message" : {
        "type" : "text"
      }
    }
  },
  "ingest_pipeline" : {
    "description" : "Ingest pipeline created by text structure finder",
    "processors" : [
      {
        "grok" : {
          "field" : "message",
          "patterns" : [
            "%{CATALINALOG}"
          ],
          "ecs_compatibility" : "v1"
        }
      },
      {
        "date" : {
          "field" : "timestamp",
          "timezone" : "{{ event.timezone }}",
          "formats" : [
            "MMM dd, yyyy h:mm:ss a"
          ]
        }
      },
      {
        "remove" : {
          "field" : "timestamp"
        }
      }
    ]
  },
  "field_stats" : {
    "java.log.origin.class.name" : {
      "count" : 7,
      "cardinality" : 6,
      "top_hits" : [
        {
          "value" : "org.apache.catalina.startup.HostConfig",
          "count" : 2
        },
        {
          "value" : "org.apache.catalina.core.AprLifecycleListener",
          "count" : 1
        },
        {
          "value" : "org.apache.catalina.core.StandardEngine",
          "count" : 1
        },
        {
          "value" : "org.apache.catalina.core.StandardService",
          "count" : 1
        },
        {
          "value" : "org.apache.catalina.startup.Catalina",
          "count" : 1
        },
        {
          "value" : "org.apache.coyote.http11.Http11Protocol",
          "count" : 1
        }
      ]
    },
    "log.level" : {
      "count" : 7,
      "cardinality" : 1,
      "top_hits" : [
        {
          "value" : "INFO",
          "count" : 7
        }
      ]
    },
    "log.origin.function" : {
      "count" : 7,
      "cardinality" : 4,
      "top_hits" : [
        {
          "value" : "deployDescriptor",
          "count" : 2
        },
        {
          "value" : "init",
          "count" : 2
        },
        {
          "value" : "start",
          "count" : 2
        },
        {
          "value" : "load",
          "count" : 1
        }
      ]
    },
    "message" : {
      "count" : 7,
      "cardinality" : 7,
      "top_hits" : [
        {
          "value" : "Deploying configuration descriptor host-manager.xml",
          "count" : 1
        },
        {
          "value" : "Deploying configuration descriptor manager.xml",
          "count" : 1
        },
        {
          "value" : "Initialization processed in 461 ms",
          "count" : 1
        },
        {
          "value" : "Initializing Coyote HTTP/1.1 on http-8080",
          "count" : 1
        },
        {
          "value" : "Starting Servlet Engine: Apache Tomcat/6.0.29",
          "count" : 1
        },
        {
          "value" : "Starting service Catalina",
          "count" : 1
        },
        {
          "value" : "The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: /usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib",
          "count" : 1
        }
      ]
    },
    "timestamp" : {
      "count" : 7,
      "cardinality" : 1,
      "earliest" : "Oct 09, 2013 12:21:37 PM",
      "latest" : "Oct 09, 2013 12:21:37 PM",
      "top_hits" : [
        {
          "value" : "Oct 09, 2013 12:21:37 PM",
          "count" : 7
        }
      ]
    }
  },
  "explanation" : [
    "Using character encoding [UTF-8], which matched the input with [15%] confidence - first [8kB] of input was pure ASCII",
    "Not NDJSON because there was a parsing exception: [[1:4] Unrecognized token 'Oct': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false') at [Source: (org.elasticsearch.xpack.textstructure.structurefinder.NdJsonTextStructureFinderFactory$ContextPrintingStringReader); line: 1, column: 4]]",
    "Not XML because there was a parsing exception: [ParseError at [row,col]:[1,1] Message: Content is not allowed in prolog.]",
    "Not CSV because 1 or more rows did not have the same number of fields as the first row (2). Bad rows [1]",
    "Not TSV because the first row has fewer than [2] fields: [1]",
    "Not semicolon delimited values because the first row has fewer than [4] fields: [1]",
    "Not vertical line delimited values because the first row has fewer than [5] fields: [1]",
    "Deciding sample is text",
    "Most likely timestamp format is [MMM dd, yyyy h:mm:ss a]",
    "A full message Grok pattern [CATALINALOG] looks appropriate"
  ]
}
```

Comparing the output of the two requests shows that Grok pattern captures are using ECS compliant naming when `v1` is specified, e.g. `class` -> `java.log.origin.class.name`, `logmessage` -> `log.function.keyword`, along with determining additional field that the legacy Grok patterns missed entirely e.g. `log.level`.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java:
```diff
@@ -213,6 +224,11 @@ public TextStructure(StreamInput in) throws IOException {
+        if (in.getVersion().onOrAfter(Version.V_8_5_0)) {
+            ecsCompatibility = getNonNullEcsCompatibilityString(in.readOptionalString());
+        } else {
+            ecsCompatibility = null;
```
Setting it to `null` here is a loophole that spoils the idea of always setting it to a valid string in the other places:

```suggestion
            ecsCompatibility = getNonNullEcsCompatibilityString(null);
```

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java:
```diff
@@ -213,6 +224,11 @@ public TextStructure(StreamInput in) throws IOException {
         shouldTrimFields = in.readOptionalBoolean();
         grokPattern = in.readOptionalString();
+        if (in.getVersion().onOrAfter(Version.V_8_5_0)) {
+            ecsCompatibility = getNonNullEcsCompatibilityString(in.readOptionalString());
```
If you remove all the loopholes that permit `null` then this doesn't need to be optional:

```suggestion
            ecsCompatibility = getNonNullEcsCompatibilityString(in.readString());
```

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java:
```diff
@@ -254,6 +270,9 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalBoolean(shouldTrimFields);
         out.writeOptionalString(grokPattern);
+        if (out.getVersion().onOrAfter(Version.V_8_5_0)) {
+            out.writeOptionalString(ecsCompatibility);
```
If you remove all the loopholes that permit `null` then this doesn't need to be optional:

```suggestion
            out.writeString(ecsCompatibility);
```

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java:
```diff
@@ -598,6 +627,11 @@ public Builder setGrokPattern(String grokPattern) {
         }
 
+        public Builder setEcsCompatibility(String ecsCompatibility) {
+            this.ecsCompatibility = TextStructure.getNonNullEcsCompatibilityString(ecsCompatibility);
```
The `Builder` class allows `ecsCompatibility` to be `null` (and defaults to that), so you might as well leave conversion of `null` to `disabled` for the constructor of the immutable object.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java:
```diff
@@ -403,6 +426,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         if (grokPattern != null && grokPattern.isEmpty() == false) {
             builder.field(GROK_PATTERN.getPreferredName(), grokPattern);
         }
+        if (ecsCompatibility != null && ecsCompatibility.isEmpty() == false) {
+            builder.field(ECS_COMPATIBILITY.getPreferredName(), ecsCompatibility);
+        }
```
If you enforce it to never be `null` or empty with `getNonNullEcsCompatibilityString()` then you don't need to check here:

```suggestion
        builder.field(ECS_COMPATIBILITY.getPreferredName(), ecsCompatibility);
```

edsavage committed:
Updated docs such that yaml rest tests now pass

Also always set the value of the ecs_compatibility request parameter in the results, regardless of the structure format, for consistency (else the default value of "disabled" will be used).

droberts195 commented on docs/reference/text-structure/apis/find-structure.asciidoc:
```diff
@@ -99,6 +99,12 @@ specified, the name of the timestamp field in the Grok pattern must match
 "timestamp". If `grok_pattern` is not specified, the structure finder creates a
 Grok pattern.
 
+`ecs_compatibility`::
+(Optional, string) The mode of compatibity with ECS compliant Grok patterns.
+When the structure finder creates a Grok pattern you can use this parameter to
+specify whether to use ECS Grok patterns instead of the legacy ones. Valid values
+are `disabled` and `v1`. The default value is `disabled`.
+
```
```suggestion
`ecs_compatibility`::
(Optional, string) The mode of compatibility with ECS compliant Grok patterns.
Use this parameter to specify whether to use ECS Grok patterns instead of
legacy ones when the structure finder creates a Grok pattern. Valid values
are `disabled` and `v1`. The default value is `disabled`. This setting primarily
has an impact when a whole message Grok pattern such as `%{CATALINALOG}`
matches the input. If the structure finder identifies a common structure but
has no idea of meaning then generic field names such as `path`, `ipaddress`,
`field1` and `field2` are used in the `grok_pattern` output, with the intention
that a user who knows the meanings rename these fields before using it.
```

droberts195 commented:
I think this is nearly ready apart from a few final tidyups.

You'll need to merge `main` into your PR branch to get CI to pass though.

edsavage committed:
Update x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java

Co-authored-by: David Roberts <dave.roberts@elastic.co>

edsavage committed:
Update x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java

Co-authored-by: David Roberts <dave.roberts@elastic.co>

edsavage committed:
Update x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java

Co-authored-by: David Roberts <dave.roberts@elastic.co>

edsavage committed:
Update x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/structurefinder/TextStructure.java

Co-authored-by: David Roberts <dave.roberts@elastic.co>

edsavage committed:
Update docs/reference/text-structure/apis/find-structure.asciidoc

Co-authored-by: David Roberts <dave.roberts@elastic.co>

edsavage committed:
Remove unnecessary string conversion

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into switch_to_ecs_grok_patterns

droberts195 reviewed (approved):
LGTM

edsavage committed:
Fixed typos in docs
